source/modulo_new_core/include/modulo_new_core/communication/MessagePairInterface.hpp

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <memory>
+#include "modulo_new_core/communication/MessageType.hpp"
+
+namespace modulo_new_core::communication {
+
+// forward declaration of derived MessagePair class
+template<typename MsgT, typename DataT>
+class MessagePair;
+
+class MessagePairInterface : public std::enable_shared_from_this<MessagePairInterface> {
+public:
+  explicit MessagePairInterface(MessageType type);
+
+  virtual ~MessagePairInterface() = default;
+
+  MessagePairInterface(const MessagePairInterface& message_pair) = default;
+
+  template<typename MsgT, typename DataT>
+  std::shared_ptr<MessagePair<MsgT, DataT>> get_message_pair(bool validate_pointer = true);
+
+  template<typename MsgT, typename DataT>
+  MsgT get_message_pair_message();
```

Comment:
For this method name I would also suggest `write_message()`.
```suggestion
  MsgT write_message();
```